File: open_raccoon/apps/raccoon_feed/lib/raccoon_feed/feed_item.ex
defmodule RaccoonFeed.FeedItem do
  @moduledoc """
  Feed item schema for the social discovery feed.
  """
  use Ecto.Schema
  import Ecto.Changeset

  @primary_key {:id, :binary_id, autogenerate: true}
  @foreign_key_type :binary_id
  schema "feed_items" do
    belongs_to :creator, RaccoonAccounts.User

    field :type, Ecto.Enum,
      values: [:agent_showcase, :page_showcase, :tool_showcase, :remix, :creation]

    field :reference_id, :binary_id
    field :reference_type, Ecto.Enum, values: [:agent, :page, :tool]
    field :title, :string
    field :description, :string
    field :thumbnail_url, :string
    field :quality_score, :float, default: 0.0
    field :trending_score, :float, default: 0.0
    field :like_count, :integer, default: 0
    field :fork_count, :integer, default: 0
    field :view_count, :integer, default: 0

    has_many :likes, RaccoonFeed.FeedLike

    timestamps(type: :utc_datetime_usec)
  end

  def changeset(feed_item, attrs) do
    feed_item
    |> cast(attrs, [
      :creator_id, :type, :reference_id, :reference_type, :title,
      :description, :thumbnail_url
    ])
    |> validate_required([:creator_id, :type, :reference_id, :reference_type])
    |> validate_length(:title, max: 255)
    |> validate_inclusion(:type, [:agent_showcase, :page_showcase, :tool_showcase, :remix, :creation])
    |> validate_inclusion(:reference_type, [:agent, :page, :tool])
    |> foreign_key_constraint(:creator_id)
  end
end

File: WaiAgents/Sources/WaiAgentsKit/Models/FeedItem.swift
import Foundation

public struct FeedItem: Codable, Identifiable, Sendable, Equatable {
    public let id: String
    public let creatorID: String
    public var type: FeedItemType
    public var referenceID: String
    public var referenceType: ReferenceType
    public var title: String?
    public var description: String?
    public var thumbnailURL: URL?
    public var qualityScore: Double
    public var trendingScore: Double
    public var likeCount: Int
    public var forkCount: Int
    public var viewCount: Int
    public let createdAt: Date
    public var updatedAt: Date

    public enum FeedItemType: String, Codable, Sendable {
        case agentShowcase = "agent_showcase"
        case pageShowcase = "page_showcase"
        case toolShowcase = "tool_showcase"
        case remix
        case creation
    }

    public enum ReferenceType: String, Codable, Sendable {
        case agent
        case page
        case tool
    }

    enum CodingKeys: String, CodingKey {
        case id
        case creatorID = "creatorId"
        case type
        case referenceID = "referenceId"
        case referenceType
        case title
        case description
        case thumbnailURL = "thumbnailUrl"
        case qualityScore
        case trendingScore
        case likeCount
        case forkCount
        case viewCount
        case createdAt
        case updatedAt
    }

    public init(
        id: String,
        creatorID: String,
        type: FeedItemType,
        referenceID: String,
        referenceType: ReferenceType,
        title: String? = nil,
        description: String? = nil,
        thumbnailURL: URL? = nil,
        qualityScore: Double = 0,
        trendingScore: Double = 0,
        likeCount: Int = 0,
        forkCount: Int = 0,
        viewCount: Int = 0,
        createdAt: Date = Date(),
        updatedAt: Date = Date()
    ) {
        self.id = id
        self.creatorID = creatorID
        self.type = type
        self.referenceID = referenceID
        self.referenceType = referenceType
        self.title = title
        self.description = description
        self.thumbnailURL = thumbnailURL
        self.qualityScore = qualityScore
        self.trendingScore = trendingScore
        self.likeCount = likeCount
        self.forkCount = forkCount
        self.viewCount = viewCount
        self.createdAt = createdAt
        self.updatedAt = updatedAt
    }
}
